### PR TITLE
add form components from ODC

### PR DESCRIPTION
**Pull request by nemesis09**
Story:
https://issues.redhat.com/browse/HAC-130

Description:
Add generic form components for HAC

Screens:
Screenshot from 2021-11-15 17-59-52

Screenshot from 2021-11-12 09-40-27

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -29,9 +29,15 @@
     "@redhat-cloud-services/frontend-components": "^3.3.3",
     "@redhat-cloud-services/frontend-components-notifications": "^3.2.2",
     "@redhat-cloud-services/frontend-components-utilities": "^3.2.4",
+    "@rjsf/core": "^3.2.0",
```
Comment (rohitkrai03) [inline package.json:32]: We don't need the dynamic forms for now.

File: src/shared/async/async.tsx [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ */
+const sameLoader = (a: () => Promise<React.ComponentType>) => (
+  b: () => Promise<React.ComponentType>,
+) => a?.name === b?.name && (a || 'a').toString() === (b || 'b').toString();
+
+enum AsyncComponentError {
+  ComponentNotFound = 'COMPONENT_NOT_FOUND',
+}
+
+export class AsyncComponent extends React.Component<AsyncComponentProps, AsyncComponentState> {
```
Comment (rohitkrai03) [inline src/shared/async/async.tsx:20]: Don't think we need to bring this. We can start to use `React.Lazy` instead.

File: src/shared/button-bar/button-bar.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
+import * as _ from 'lodash-es';
```
Comment (rohitkrai03) [inline src/shared/button-bar/button-bar.tsx:1]: Components should be named in `PascalCase`.

Comment (rohitkrai03) [reply src/shared/button-bar/button-bar.tsx:1]: You should also create a component folder inside `shared`.

File: src/shared/button-bar/button-bar.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        {errorMessage && <ErrorMessage message={errorMessage} />}
+        {injectDisabled(children, inProgress)}
+        {inProgress && <LoadingInline />}
+        {infoMessage && <InfoMessage message={infoMessage} />}
+      </AlertGroup>
+    </div>
+  );
+};
+
+ButtonBar.propTypes = {
```
Comment (rohitkrai03) [inline src/shared/button-bar/button-bar.tsx:71]: Why are you using `proptypes` in typescript?

File: src/shared/dynamic-form/fields.tsx [hunk 5]
```diff
@@ -0,0 +1,357 @@
+import * as React from 'react';
```
Comment (rohitkrai03) [inline src/shared/dynamic-form/fields.tsx:1]: No need to bring in dynamic form stuff.

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]
Why is your PR removing almost all of `package-lock.json`?

File: src/shared/button-bar/ButtonBar.tsx [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        {errorMessage && <ErrorMessage message={errorMessage} />}
+        {injectDisabled(children, inProgress)}
+        {inProgress && <Loading inline />}
+        {infoMessage && <InfoMessage message={infoMessage} />}
+      </AlertGroup>
+    </div>
+  );
+};
+
+ButtonBar.propTypes = {
```
Comment (rohitkrai03) [inline src/shared/button-bar/ButtonBar.tsx:53]: Why are you using proptypes?

File: src/pages/SampleForm/SampleFormPage.tsx [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+import React from 'react';
+import { Main } from '@redhat-cloud-services/frontend-components/Main';
+import { PageHeader, PageHeaderTitle } from '@redhat-cloud-services/frontend-components/PageHeader';
+import { Formik, FormikProps } from 'formik';
+
+const SampleForm = React.lazy(() => import('../../components/SampleForm/SampleForm'));
+
+const SampleFormPage: React.FC = () => {
+  React.useEffect(() => {
+    window.insights?.chrome?.appAction?.('sample-form');
```
Comment (rohitkrai03) [inline src/pages/SampleForm/SampleFormPage.tsx:12]: Why do you need this here?

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

**Comment by rohitkrai03** [comment]
Take a look at why the tests are failing.

**Comment by rohitkrai03** [comment]
@nemesis09 You have lint errors now.

**Comment by rohitkrai03** [comment]
@nemesis09 You still have lint errors and test failures.

File: src/Routes.tsx [hunk 8]
```diff
@@ -24,7 +27,8 @@ export const Routes: React.FC = () => (
     }
   >
     <Switch>
-      <Route path="/" component={SamplePage} />
+      <Route path="/" component={SampleFormPage} />
+      <Route path="/sample-form" component={SamplePage} />
```
Comment (rohitkrai03) [inline src/Routes.tsx:31]:
```suggestion
      <Route path="/" component={SamplePage} />
      <Route path="/sample-form" component={SampleFormPage} />
```

Comment (rohitkrai03) [reply src/Routes.tsx:31]: Add a link to the sample page for the form page.

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]
Overall changes look good. Just a few chnages needed - 
 - The select input field doesn't work.
 - There are some CSS issues in multi column field.

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]
@nemesis09 You have issues with package lock file again.

File: src/shared/components/index.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+export * from './button-bar';
+export * from './close-button';
+export * from './drag-drop';
+export * from './dropdown';
+export * from './form-components';
+export * from './formik-fields';
+export * from './loading';
```
Comment (karthikjeeyar) [inline src/shared/components/index.ts:7]: Your last commit 7057de8 removed the loading folder
```suggestion

```

**Comment by karthikjeeyar** [comment]
Looks good overall, I observed the below issue
While selecting a new option in SelectInputField, it splits the preselected option into individual characters.

https://user-images.githubusercontent.com/9964343/141932744-cbecfd17-e5db-4db0-afc6-87cd8b95a745.mov

@nemesis09  should we bring in environmentField as well ?

File: src/shared/components/index.ts (hunk 9 above)
Comment (nemesis09) [reply src/shared/components/index.ts]: fixed

**Comment by nemesis09** [comment]
> Looks good overall, I observed the below issue While selecting a new option in SelectInputField, it splits the preselected option into individual characters.
> Screen.Recording.2021-11-16.at.11.49.42.AM.mov
> 
> @nemesis09 should we bring in environmentField as well ?

fixed the select field.
as discussed,  this PR only brings basic components. we can have a separate PR for other fields

File: src/shared/components/form-components/FormFooter.scss [hunk 10]
```diff
@@ -0,0 +1,14 @@
+.hacDev-form-footer {
+  padding: var(--pf-global--spacer--md) (--pf-global--gutter/2);
+  @media (min-width: 768px) {
+    padding: var(--pf-global--spacer--md) (--pf-global--gutter);
+  }
```
Comment (jerolimov) [inline src/shared/components/form-components/FormFooter.scss:5]: I don't know if this is the issue, but the sticky footer doesn't look nice at the moment:

footer

I didn't find a quick-fix solution. We can merge this now, but then we should create a follow-up bug to solve this next sprint. Wdyt?

**Review by jerolimov — CHANGES_REQUESTED** [review]
@nemesis09 And is it possible that we also add all existing unit tests from console?

See `packages/console-shared/src/components/form-utils/__tests__/*`, maybe there are more folders.

File: src/shared/components/form-components/FormFooter.scss (hunk 10 above)
Comment (nemesis09) [reply src/shared/components/form-components/FormFooter.scss:5]: I agree, let us merge this for now. I think this is what we have in console. But we can try to make it better with a follow up PR if we want.

**Comment by nemesis09** [comment]
> @nemesis09 And is it possible that we also add all existing unit tests from console?
> 
> See `packages/console-shared/src/components/form-utils/__tests__/*`, maybe there are more folders.

I'll have a separate PR for the tests later.

File: src/shared/components/form-components/FlexForm.tsx [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
+import { Form, FormProps } from '@patternfly/react-core';
+
+// type FlexFormProps = {
+//   children?: React.ReactNode;
+// };
+
```
Comment (jerolimov) [inline src/shared/components/form-components/FlexForm.tsx:7]: We should avoid commented code, just remove it

```suggestion
```

**Review by jerolimov — APPROVED** [review]
Has lgtm by Karthik before, looks fine to me as well. But we should definitely add unit tests asap.

/lgtm
/approve

**Pull request merged**